libslide/src/grammar/intern.rs

```diff
@@ -131,15 +204,36 @@ macro_rules! make_interner {
         impl Deref for $interned_struct {
             type Target = <Self as InternedExpression>::Inner;
 
-            fn deref(&self) -> &<Self as InternedExpression>::Inner {
-                let table = $spur2expr.read().expect("Failed to read intern reverse map; likely poisoned.");
-                let val = table.get(&self.0).expect("BUG: intern key does not map to an interned value");
+            fn deref(&self) -> &Self::Target {
+                let interner_lk = $interner.read().expect("Failed to read interner; likely poisoned.");
+                let interned_p = interner_lk.spur_map.get(&(self.spur, self.span)).unwrap_or_else(|| panic!(
+                    // Constructing this formatted string is decently expensive, so avoid it unless
+                    // we actually can't find the interned value.
+                    "BUG: intern key does not map to an interned value.
+                    Self: {:?}
+                    Rodeo: {:?}
+                    Spur Map: {:?}", self, interner_lk.rodeo, interner_lk.spur_map
+                ));
+                let inner: &Self::Target = &**interned_p;
                 // Safety:
                 // - semantics: OK because cast to ptr, then deref ptr and return reference
                 // - lifetimes: OK because values present in the intern tables last for the lifetime
                 //   of the table, which is static.
+                // - invalidation of hashmap reference: OK because we are returning a reference to
+                //   the expression being pointed at by the reference, not the reference itself. As
+                //   an illustration, suppose our map looks like
+                //
+                //   |   |   |   | k |   |     (before reallocation)
+                //   |   |   |   | v |   |
+                //                 \
+                //          /-------\-> expr -- return a reference to this
+                //         /
+                //   |   | v |   |   |   |
+                //   |   | k |   |   |   |     (after reallocation)
+                //
+                //   so even if the map is reallocated, our reference is valid.
                 unsafe {
```

Comment:
can you elaborate just for me. I don't understand what you mean by we are returning a ref to an expression. Does that just mean dereferencing? If it is, then i overthought.

Reply:
Updated the comment, lmk what you think.